Libplanet.Tests/Action/ActionEvaluatorTest.cs

```diff
@@ -0,0 +1,997 @@
+                transactions: txs).AttachStateRootHash(stateStore, null);
+            StateGetter<RandomAction> nullStateGetter =
+                (address, hashDigest, stateCompleter) => null;
+            BalanceGetter<RandomAction> nullBalanceGetter =
+                (address, currency, hashDigest, fungibleAssetStateCompleter)
+                    => new FungibleAssetValue(currency);
+            var actionEvaluator =
+                new ActionEvaluator<RandomAction>(
+                    policyBlockAction: null,
+                    stateGetter: nullStateGetter,
+                    balanceGetter: nullBalanceGetter,
+                    trieGetter: null);
+            var generatedRandomNumbers = new List<int>();
+
+            Assert.NotEqual(stateRootBlock.Hash, noStateRootBlock.Hash);
+            Assert.Equal(stateRootBlock.PreEvaluationHash, noStateRootBlock.PreEvaluationHash);
+
+            for (int i = 0; i < repeatCount; ++i)
+            {
+                var actionEvaluations = actionEvaluator.Evaluate(
+                    noStateRootBlock,
+                    StateCompleterSet<RandomAction>.Reject);
+                generatedRandomNumbers.Add(
+                    (Integer)actionEvaluations[0].OutputStates.GetState(txAddress));
+                actionEvaluations = actionEvaluator.Evaluate(
+                    stateRootBlock,
+                    StateCompleterSet<RandomAction>.Reject);
+                generatedRandomNumbers.Add(
+                    (Integer)actionEvaluations[0].OutputStates.GetState(txAddress));
+            }
+
+            for (int i = 1; i < generatedRandomNumbers.Count; ++i)
+            {
+                Assert.Equal(generatedRandomNumbers[0], generatedRandomNumbers[i]);
+            }
+        }
+
+        [Fact]
+        public async void Evaluate()
+        {
+            var privateKey = new PrivateKey();
+            var address = privateKey.ToAddress();
+            long blockIndex = 1;
+
+            TestEvaluateAction action = new TestEvaluateAction();
+
+            var store = new DefaultStore(null);
+            var stateStore =
+                new TrieStateStore(new MemoryKeyValueStore(), new MemoryKeyValueStore());
+            var chain = TestUtils.MakeBlockChain<TestEvaluateAction>(
+                policy: new BlockPolicy<TestEvaluateAction>(),
+                store: store,
+                stateStore: stateStore);
+            var tx = Transaction<TestEvaluateAction>.Create(
+                nonce: 0,
+                privateKey: privateKey,
+                genesisHash: chain.Genesis.Hash,
+                actions: new[] { action });
+
+            chain.StageTransaction(tx);
+            await chain.MineBlock(_storeFx.Address1);
+
+            var actionEvaluation = chain.ActionEvaluator.Evaluate(
+                chain.Tip,
+                StateCompleterSet<TestEvaluateAction>.Recalculate);
+            Assert.False(actionEvaluation[0].InputContext.BlockAction);
```

Comment:
Could you test if `ActionEvaluation`'s other properties have correct values too?  Especially, it would be better if it properly has `Exception` on failed actions.

Reply:
@greymistcube It seems this is not addressed yet.

Reply:
Still working on it. 😢

Reply:
Implemented with mostly copy pasted code. See 55e4ec57c71953cead85c364719d3832aa5fad56.